pkg/reconciler/revision/resources/deploy.go

```diff
@@ -136,49 +135,63 @@ func makePodSpec(rev *v1.Revision, loggingConfig *logging.Config, tracingConfig
 	return podSpec, nil
 }
 
-// BuildUserContainer makes a container from the Revision template.
-func BuildUserContainer(rev *v1.Revision) *corev1.Container {
-	userContainer := rev.Spec.GetContainer().DeepCopy()
+// BuildUserContainers makes an array of containers from the Revision template.
+func BuildUserContainers(rev *v1.Revision) []corev1.Container {
+	containers := make([]corev1.Container, 0, len(rev.Spec.PodSpec.Containers))
+	for i := range rev.Spec.PodSpec.Containers {
+		var container corev1.Container
+		if len(rev.Spec.PodSpec.Containers[i].Ports) != 0 || len(rev.Spec.PodSpec.Containers) == 1 {
+			container = makeServingContainer(*rev.Spec.PodSpec.Containers[i].DeepCopy(), rev)
+		} else {
+			container = makeContainer(*rev.Spec.PodSpec.Containers[i].DeepCopy(), rev)
+		}
+		if rev.Status.ContainerStatuses != nil {
```

Comment:
any reason not to avoid differentiating between nil and empty array?

```suggestion
		if len(rev.Status.ContainerStatuses) != 0 {
```

Reply:
IMO for `rev.Status.ContainerStatuses` i did not see any difference between nil and empty array
Because it works with both checks 

I see Main difference comes when we declare slice like below

```
var a []string
```
```
a := []string{}
```
Let me know your opinion on this

Reply:
it's not a huge deal, it's just that if the check is for nil rather than len(0) then the interface to this method differentiates between nil and empty array, which can lead to hard to debug problems (see [this link](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices) "when designing interfaces, avoid making a distinction between a nil slice and a non-nil, zero-length slice, as this can lead to subtle programming errors" and [this talk](https://www.youtube.com/watch?v=ynoY2xz-F8s)). It means you have to be careful not to pass this method a revision with an empty rather than nil ContainerStatus array, which is a little unexpected, since generally in go empty and nil arrays should be interchangeable.

Since it's equally clear either way it seems nicer to avoid the gotcha.